python/pip_install/tools/wheel_installer/wheel.py

```diff
@@ -228,14 +234,27 @@ class Deps:
     def __init__(
         self,
         name: str,
+        *,
+        requires_dist: Optional[List[str]],
         extras: Optional[Set[str]] = None,
         platforms: Optional[Set[Platform]] = None,
     ):
         self.name: str = Deps._normalize(name)
+        self._platforms: Set[Platform] = platforms or set()
+
+        reqs = sorted(
+            [Requirement(wheel_req) for wheel_req in requires_dist],
+            key=lambda x: f"{x.name}:{sorted(x.extras)}",
+        )
+
+        # Resolve any extra extras due to self-edges, empty string means not extras
```

Comment:
I think I commented on the empty string thing in an earlier PR; this empty string thing is confusing.

Does the presence of the empty string mean no extras? i.e. are `{"", "foo"}` Or is it assumed that an empty string element is mutually exclusive with other elements? Why is "no extras" communicated using set-with-only-empty-string as opposed to empty-set ?

(I've authored and had to maintain code where None vs empty set vs non-empty set have 3 different meanings, and it was always a cognitive burden)

Reply:
Right now the case of `{"", "foo"}` and `{"foo"}` has the same meaning. I moved this detail into the method implementation to better isolate it from the rest of the code as it is an implementation detail. The return value of `_resolve_extras` should be always a set with at least one element in it and it just makes the matching with `any(...)` easier to write.